// datocms-client/lib/local/fields/File.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _queryString = require('query-string');

var _queryString2 = _interopRequireDefault(_queryString);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var File = function () {
  function File(value, _ref) {
    var itemsRepo = _ref.itemsRepo;
    var _ref$imgixHost = _ref.imgixHost;
    var imgixHost = _ref$imgixHost === undefined ? 'https://www.datocms-assets.com' : _ref$imgixHost;

    _classCallCheck(this, File);

    this.imgixHost = imgixHost;
    this.itemsRepo = itemsRepo;
    this.upload = itemsRepo.entitiesRepo.findEntity('upload', value);
  }

  _createClass(File, [{
    key: 'url',
    value: function url() {
      var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      if (params && Object.keys(params).length > 0) {
        return '' + this.imgixHost + this.path + '?' + _queryString2.default.stringify(params);
      }
      return '' + this.imgixHost + this.path;
    }
  }, {
    key: 'toMap',
    value: function toMap() {
      return {
        format: this.format,
        size: this.size,
        width: this.width,
        height: this.height,
        title: this.title,
        alt: this.alt,
        url: this.url()
      };
    }
  }, {
    key: 'path',
    get: function get() {
      return this.upload.path;
    }
  }, {
    key: 'format',
    get: function get() {
      return this.upload.format;
    }
  }, {
    key: 'size',
    get: function get() {
      return this.upload.size;
    }
  }, {
    key: 'width',
    get: function get() {
      return this.upload.width;
    }
  }, {
    key: 'height',
    get: function get() {
      return this.upload.height;
    }
  }, {
    key: 'alt',
    get: function get() {
      return this.upload.alt;
    }
  }, {
    key: 'title',
    get: function get() {
      return this.upload.title;
    }
  }]);

  return File;
}();

exports.default = File;
module.exports = exports['default'];
