// datocms-client/lib/site/repos/AccessTokenRepo.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _deserializeJsonApi = require('../../deserializeJsonApi');

var _deserializeJsonApi2 = _interopRequireDefault(_deserializeJsonApi);

var _serializeJsonApi = require('../../serializeJsonApi');

var _serializeJsonApi2 = _interopRequireDefault(_serializeJsonApi);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var AccessTokenRepo = function () {
  function AccessTokenRepo(client) {
    _classCallCheck(this, AccessTokenRepo);

    this.client = client;
  }

  _createClass(AccessTokenRepo, [{
    key: 'create',
    value: function create(resourceAttributes) {
      var serializedResource = (0, _serializeJsonApi2.default)(resourceAttributes, {
        type: 'access_token',
        attributes: ['name'],
        requiredAttributes: ['name'],
        relationships: {
          role: 'role'
        }
      });
      return this.client.post('/access_tokens', serializedResource).then(function (response) {
        return Promise.resolve((0, _deserializeJsonApi2.default)(response));
      });
    }
  }, {
    key: 'update',
    value: function update(userId, resourceAttributes) {
      var serializedResource = (0, _serializeJsonApi2.default)(userId, resourceAttributes, {
        type: 'access_token',
        attributes: ['name'],
        relationships: {
          role: 'role'
        }
      });
      return this.client.put('/access_tokens/' + userId, serializedResource).then(function (response) {
        return Promise.resolve((0, _deserializeJsonApi2.default)(response));
      });
    }
  }, {
    key: 'all',
    value: function all() {
      return this.client.get('/access_tokens').then(function (response) {
        return Promise.resolve((0, _deserializeJsonApi2.default)(response));
      });
    }
  }, {
    key: 'find',
    value: function find(userId) {
      return this.client.get('/access_tokens/' + userId).then(function (response) {
        return Promise.resolve((0, _deserializeJsonApi2.default)(response));
      });
    }
  }, {
    key: 'regenerateToken',
    value: function regenerateToken(userId) {
      return this.client.post('/access_tokens/' + userId + '/regenerate_token').then(function (response) {
        return Promise.resolve((0, _deserializeJsonApi2.default)(response));
      });
    }
  }, {
    key: 'destroy',
    value: function destroy(userId) {
      return this.client.delete('/access_tokens/' + userId).then(function (response) {
        return Promise.resolve((0, _deserializeJsonApi2.default)(response));
      });
    }
  }]);

  return AccessTokenRepo;
}();

exports.default = AccessTokenRepo;
module.exports = exports['default'];
