// datocms-client/lib/site/repos/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _DeployEventRepo = require('./DeployEventRepo');

Object.defineProperty(exports, 'deployEvents', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_DeployEventRepo).default;
  }
});

var _FieldRepo = require('./FieldRepo');

Object.defineProperty(exports, 'fields', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_FieldRepo).default;
  }
});

var _ItemRepo = require('./ItemRepo');

Object.defineProperty(exports, 'items', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ItemRepo).default;
  }
});

var _ItemTypeRepo = require('./ItemTypeRepo');

Object.defineProperty(exports, 'itemTypes', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ItemTypeRepo).default;
  }
});

var _MenuItemRepo = require('./MenuItemRepo');

Object.defineProperty(exports, 'menuItems', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_MenuItemRepo).default;
  }
});

var _SiteRepo = require('./SiteRepo');

Object.defineProperty(exports, 'site', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_SiteRepo).default;
  }
});

var _UploadRequestRepo = require('./UploadRequestRepo');

Object.defineProperty(exports, 'uploadRequests', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_UploadRequestRepo).default;
  }
});

var _UserRepo = require('./UserRepo');

Object.defineProperty(exports, 'users', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_UserRepo).default;
  }
});

var _SearchResultRepo = require('./SearchResultRepo');

Object.defineProperty(exports, 'searchResults', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_SearchResultRepo).default;
  }
});

var _RoleRepo = require('./RoleRepo');

Object.defineProperty(exports, 'roles', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_RoleRepo).default;
  }
});

var _AccessTokenRepo = require('./AccessTokenRepo');

Object.defineProperty(exports, 'accessTokens', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_AccessTokenRepo).default;
  }
});

var _UploadRepo = require('./UploadRepo');

Object.defineProperty(exports, 'uploads', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_UploadRepo).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
